manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,289 @@
+	events := make([]api.Event, len(resourceDeallocators)+2)
+	// we want to know when tasks are deleted by the reaper
+	events[0] = api.EventDeleteTask{}
+	// and when services get marked for deletion
+	events[1] = api.EventUpdateService{}
+	i := 2
+	for _, resourceDeallocator := range resourceDeallocators {
+		events[i] = resourceDeallocator.event
+		i++
+	}
+
+	var allServices []*api.Service
+	resourceLists := make(map[reflect.Type][]serviceLevelResource)
+
+	eventsChan, _, err := store.ViewAndWatch(deallocator.store,
+		func(readTx store.ReadTx) (err error) {
+			// look for services that are marked for deletion
+			// there's no index on the `PendingDelete` field,
+			// so we just iterate over all of them and filter manually
+			// this is okay since we only do this at leadership change
+			allServices, err = store.FindServices(readTx, store.All)
+
+			if err != nil {
+				log.G(ctx).WithError(err).Error("failed to list services in deallocator init")
+				return err
+			}
+
+			// now we also need to look at all existing service-level resources
+			// that are marked for deletion
+			for key, resourceDeallocator := range resourceDeallocators {
+				allResources, err := resourceDeallocator.enumerator(readTx)
+
+				if err != nil {
+					log.G(ctx).WithError(err).Errorf("failed to list %vs in deallocator init", resourceDeallocator.name)
+					return err
+				}
+
+				resourceLists[key] = allResources
+			}
+
+			return
+		},
+		events...)
+
+	if err != nil {
+		// if we have an error here, we can't proceed any further, let the world burn
+		log.G(ctx).WithError(err).Error("failed to initialize the deallocator")
+		return err
+	}
+
+	defer func() {
+		// eventsChanCancel()
+		close(deallocator.doneChan)
+	}()
+
+	anyUpdated := false
+	// now let's populate our internal taskCounts
+	for _, service := range allServices {
+		if updated, _ := deallocator.processService(ctx, service); updated {
+			anyUpdated = true
+		}
+	}
+
+	// and deallocate resources that are pending for deletion and aren't used any more
+	for key, resourceList := range resourceLists {
+		resourceDeallocator := resourceDeallocators[key]
+
+		for _, resource := range resourceList {
+			if updated, _ := deallocator.processResource(ctx, nil, resourceDeallocator, resource, nil); updated {
+				anyUpdated = true
+			}
+		}
+	}
+
+	// now we just need to wait for events
+	deallocator.notifyEventChan(anyUpdated)
+	for {
+		select {
+		case event := <-eventsChan:
+			if updated, err := deallocator.processNewEvent(ctx, event); err == nil {
+				deallocator.notifyEventChan(updated)
+			} else {
+				log.G(ctx).WithError(err).Errorf("error processing deallocator event %#v", event)
+			}
+		case <-deallocator.stopChan:
+			return nil
+		}
+	}
+}
+
+// Stop stops the deallocator's routine
+// FIXME (jrouge): see the comment on TaskReaper.Stop() and see when to properly stop this
+// plus unit test on this!
+func (deallocator *Deallocator) Stop() {
+	close(deallocator.stopChan)
```

Comment:
you want to wrap this in a `sync.Once` to make calls to `Stop` idempotent.

Reply:
I'll do that when I get around to actually using it, right now I'm not clear on this. That's what the comment right before it hints at. But good catch, keeping it in mind :)